pkg/snowflake/table.go

```diff
@@ -0,0 +1,144 @@
+package snowflake
+
+import (
+	"database/sql"
+	"fmt"
+	"strings"
+
+	"github.com/jmoiron/sqlx"
+)
+
+// TableBuilder abstracts the creation of SQL queries for a Snowflake schema
+type TableBuilder struct {
+	name    string
+	db      string
+	schema  string
+	columns []map[string]string
+	comment string
+}
+
+// QualifiedName prepends the db and schema if set and escapes everything nicely
+func (tb *TableBuilder) QualifiedName() string {
+	var n strings.Builder
+
+	if tb.db != "" && tb.schema != "" {
+		n.WriteString(fmt.Sprintf(`"%v"."%v".`, tb.db, tb.schema))
+	}
+
+	if tb.db != "" && tb.schema == "" {
+		n.WriteString(fmt.Sprintf(`"%v"..`, tb.db))
+	}
+
+	if tb.db == "" && tb.schema != "" {
+		n.WriteString(fmt.Sprintf(`"%v".`, tb.schema))
+	}
+
+	n.WriteString(fmt.Sprintf(`"%v"`, tb.name))
+
+	return n.String()
+}
+
+// WithComment adds a comment to the TableBuilder
+func (tb *TableBuilder) WithComment(c string) *TableBuilder {
+	tb.comment = c
+	return tb
+}
+
+// WithColumns sets the column definitions on the TableBuilder
+func (tb *TableBuilder) WithColumns(c []map[string]string) *TableBuilder {
+	tb.columns = c
+	return tb
+}
+
+// Table returns a pointer to a Builder that abstracts the DDL operations for a table.
+//
+// Supported DDL operations are:
+//   - ALTER TABLE
+//   - DROP TABLE
+//   - SHOW TABLES
+//
+// [Snowflake Reference](https://docs.snowflake.com/en/sql-reference/ddl-table.html)
+func Table(name, db, schema string) *TableBuilder {
+	return &TableBuilder{
+		name:   name,
+		db:     db,
+		schema: schema,
+	}
+}
+
+// Table returns a pointer to a Builder that abstracts the DDL operations for a table.
+//
+// Supported DDL operations are:
+//   - CREATE TABLE
+//
+// [Snowflake Reference](https://docs.snowflake.com/en/sql-reference/ddl-table.html)
+func TableWithColumnDefinitions(name, db, schema string, columns []map[string]string) *TableBuilder {
+	return &TableBuilder{
+		name:    name,
+		db:      db,
+		schema:  schema,
+		columns: columns,
+	}
+}
+
+// Create returns the SQL statement required to create a table
+func (tb *TableBuilder) Create() string {
+	q := strings.Builder{}
+	q.WriteString(fmt.Sprintf(`CREATE TABLE %v`, tb.QualifiedName()))
+
+	q.WriteString(fmt.Sprintf(` (`))
+	columnDefinitions := []string{}
+	for _, columnDefinition := range tb.columns {
+		columnDefinitions = append(columnDefinitions, fmt.Sprintf(`"%v" %v`, columnDefinition["name"], columnDefinition["type"]))
```

Comment:
I think we either need to escape or validate the contents of column definitions. Otherwise we are open to injections.

Reply:
I've just spent time rewriting sections to make use of the bind arguments feature, to allow the SQL engine to handle escaping neatly for me, only to find that as far as I can tell you can't use it to bind column names in statements.
https://docs.snowflake.com/en/sql-reference/stored-procedures-usage.html#binding-variables
https://stackoverflow.com/questions/63025655/trying-to-pass-parameter-as-binding-variable-in-snowflake-statement

I'm unsure how best to approach adding this in, would you happen to have any suggestions of packages or approaches to look at @ryanking ? Thanks!

Reply:
Yes, binding arguments does not work at all with the go snowflake driver, so we do the escaping work ourselves.

Reply:
Ah, sorry @ryanking , think I was understanding at cross-purposes to what you meant. I've amended this to also run the column defs through the existing escaping function, to make it consistent with the other resources. Is that the desired result for you? Thanks!